cartographer/sensor/point_cloud.h

```diff
@@ -54,6 +54,35 @@ class PointCloud {
 
   void push_back(PointType value);
 
+  // Creates a PointCloud consisting of all the points for which `predicate`
+  // returns true, together with the corresponding intensities.
+  template <class UnaryPredicate>
+  PointCloud copy_if(UnaryPredicate predicate) const {
+    std::vector<PointType> points;
+    std::vector<float> intensities;
+
+    // Note: benchmarks show that it is better to have this conditional outside
+    // the loop.
```

Comment:
Makes sense. Out of curiosity I was thinking if sth like this would also work, but with your explicit two loops we are on the safe side.

```cpp
      const bool with_intensities = !intensities_.empty();
      for (size_t index = 0; index < size(); ++index) {
        const PointType& point = points_[index];
        if (predicate(point)) {
          points.push_back(point);
          if (with_intensities) {
            intensities.push_back(intensities_[index]);
          }
        }
      }
```